File: src/307. Range Sum Query - Mutable.js
export class NumArray {
  /**
   * @param {number[]} nums
   */
  constructor(nums) {
    this.nums = Array(nums.length).fill(0)
    this.sums = Array(nums.length + 1).fill(0)
    for (let i = 0; i < nums.length; i++) {
      this.update(i, nums[i])
    }
  }
  /**
   * @param {number} index
   * @param {number} val
   * @return {void}
   */
  update(index, val) {
    const delta = val - this.nums[index]
    for (let i = index + 1; i < this.sums.length; i += this.lowbit(i)) {
      this.sums[i] += delta
    }
    this.nums[index] = val
  }
  /**
   * @param {number} left
   * @param {number} right
   * @return {number}
   */
  sumRange(left, right) {
    return this.getSum(right + 1) - this.getSum(left)
  }
  getSum(index) {
    let sum = 0
    for (let i = index; i > 0; i -= this.lowbit(i)) {
      sum += this.sums[i]
    }
    return sum
  }
  lowbit(x) {
    return x & -x
  }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * var obj = new NumArray(nums)
 * obj.update(index,val)
 * var param_2 = obj.sumRange(left,right)
 */

File: src/307. Range Sum Query - Mutable.cpp
class NumArray {
private:
    vector<int> sums;
    vector<int> arr;
    int lowbit (int x) {
        return x & (-x);
    }
    int getSum (int index) {
        int sum = 0;
        for (int i = index; i > 0; i -= lowbit(i)) {
            sum += sums[i];
        }
        return sum;
    }
public:
    NumArray(vector<int>& nums) {
        int n = nums.size();
        sums.resize(n + 1);
        arr.resize(n + 1);
        for (int i = 0; i < n; ++i) {
            update(i, nums[i]);
        }
    }
    
    void update(int index, int val) {
        int delta = val - arr[index];
        for (int i = index + 1; i < sums.size(); i += lowbit(i)) {
            sums[i] += delta;
        }
        arr[index] += delta;
    }
    
    int sumRange(int left, int right) {
        return getSum(right + 1) - getSum(left);
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(index,val);
 * int param_2 = obj->sumRange(left,right);
 */